core/client/fs/src/main/java/alluxio/client/file/FileInStream.java

```diff
@@ -468,21 +485,56 @@ private void handleCacheStreamException(IOException e) {
   }
 
   /**
-   * Only updates {@link #mCurrentCacheStream}, {@link #mCurrentBlockInStream} and
-   * {@link #mStreamBlockId} to be in sync with the current block (i.e. {@link #getCurrentBlockId()}
-   * ). If this method is called multiple times, the subsequent invokes become no-op. Call this
-   * function every read and seek unless you are sure about the block streams are up-to-date.
+   * Checks whether the current block stream has remaining bytes, if not, advances to the next block
+   * stream. Otherwise, this is a no-op.
+   *
+   * This should only be used in streaming read APIs. For APIs like {@link #seek(long)},
+   * {@link #skip(long)}, the block stream should be updated according to the current read position,
+   * use {@link #updateStreamsOnSeek()} instead.
+   */
+  private void updateStreamsOnRead() throws IOException {
+    checkCacheStreamInSync();
+    if (mBlockIndex == UNINITIALIZED_BLOCK_INDEX) {
+      // Initializes the streams for the first block.
+      mBlockIndex = 0;
+      updateStreamsInternal();
+      return;
+    }
+    if (shouldUpdateStreams()) {
+      mBlockIndex++;
+      updateStreamsInternal();
+    }
+  }
+
+  /**
+   * Computes the current block stream ID based on the current read position, if the stream ID is
+   * different from the current block stream ID, or the stream ID is the same as the current
+   * block stream ID but the current stream has no remaining bytes, then updates the stream.
+   *
+   * Comparing to {@link #updateStreamsOnRead()}, this method has overhead for computing the
+   * block stream ID, only use this method for non-streaming read APIs like {@link #seek(long)}
+   * and {@link #skip(long)}.
+   */
+  private void updateStreamsOnSeek() throws IOException {
+    checkCacheStreamInSync();
+    int blockIndex = (int) (mPos / mBlockSize);
```

Comment:
Optimization opportunity:

In construction:
```java
// if the block size is power of two, which is the common case.
mIsBlockSizePowerOfTwo = mBlockSize & (mBlockSize - 1) == 0; 
if (mIsBlockSizePowerOfTwo) {
  // if we want to go crazy, check out https://stackoverflow.com/questions/21442088/computing-the-floor-of-log-2x-using-only-bitwise-operators-in-c
  mBlockSizeBits = (int)(Math.log(mBlockSize)/Math.log(2); // log2(mBlockSize)
}
```
Then in `updateStreamsOnSeek()`, we can do:
```java
blockIndex = mIsBlockSizePowerOfTwo ? mPos >> mBlockSizeBits : ((int) (mPos / mBlockSize))
```
to convert expensive division to really cheap right-shift bit-ops.

This can apply to other `(int) (mPos / mBlockSize)` use in this file.

Reply:
make sense

Reply:
Is division that much more expensive than shift operations (I think its less than 2x)? I think the best case is if we can get the JIT to optimize the division because the block size is a constant.

Reply:
division is much more expensive than bit-shifts. However, I just recall this is Java and JIT is almost certain doing this optimization for you @uronce-cc . If we don't see non-trivial benefits here, let's still use division to keep readability.